pgsql/li_editor.c

```diff
@@ -465,3 +465,53 @@ Datum lipoint_projective(PG_FUNCTION_ARGS)
 
 	PG_RETURN_POINTER(serpoint);
 }
+
+/**
+* Apply an affine transformation to a box4d
+* PC_Affine(box libox4d, mat33 float8[9], off float8[3]) returns libox4d
+*/
+PG_FUNCTION_INFO_V1(libox4d_affine);
+Datum libox4d_affine(PG_FUNCTION_ARGS)
+{
+	int num_elts;
+	LIBOX4D *ibox, *obox;
+	ArrayType *mat33, *vec3;
+	float8 *mat33_values, *vec3_values;
+
+	if ( PG_ARGISNULL(0) )
+		PG_RETURN_NULL();	/* returns null if no input values */
+
+	ibox = (LIBOX4D *) PG_GETARG_POINTER(0);
+
+	mat33 = PG_GETARG_ARRAYTYPE_P(1);
+	if ( ARR_ELEMTYPE(mat33) != FLOAT8OID )
+		elog(ERROR, "mat33 must be of type float8[]");
+	if ( ARR_NDIM(mat33) != 1 )
+		elog(ERROR, "mat33 array must have only one dimension");
+	if ( ARR_HASNULL(mat33) )
+		elog(ERROR, "mat33 array must not have null elements");
+	num_elts = ARR_DIMS(mat33)[0];
+	if ( num_elts != 9 )
+		elog(ERROR, "mat33 array must have 9 elements");
+	mat33_values = (float8 *) ARR_DATA_PTR(mat33);
+
+	vec3 = PG_GETARG_ARRAYTYPE_P(2);
+	if ( ARR_ELEMTYPE(vec3) != FLOAT8OID )
+		elog(ERROR, "vec3 must be of type float8[]");
+	if ( ARR_NDIM(vec3) != 1 )
+		elog(ERROR, "vec3 array must have only one dimension");
+	if ( ARR_HASNULL(vec3) )
+		elog(ERROR, "vec3 array must not have null elements");
+	num_elts = ARR_DIMS(vec3)[0];
+	if ( num_elts != 3 )
+		elog(ERROR, "vec3 array must have 3 elements");
+	vec3_values = (float8 *) ARR_DATA_PTR(vec3);
```

Comment:
I expect many float8[] unpacking like these 2. Let us move that to a utility function